Add Support For Symbolic Shapes in Register_replacement, SDPA Pattern Matching

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115441
* #116158

Many of our pattern matching replacements are specified as a `search_fn` and a `replacment_fn`. The search_fn's are traced out once with static shapes, converted to a pattern, and then matched on every graph compiled with inductor. 
 
The static shape patterns would not match with graphs that are traced out with dynamic shapes because SymInts would be added to the graph as `sym_size` fx nodes which added additional uses and prevented matching. The previous PR partially addresses this by deduping SymInts that are resolvable to graph inputs, as is the calling convention in aot autograd. 

This PR adjusts our matching of the `search_fn` by adding SymInts to the arguments we trace out the search_fn with so that their symint accesses are deduped. Later, if we have a match, we will trace out the replacement graph with the correct Tensors and corresponding symbolic shapes that will get added to the graph.

Note: the replacement patterns will insert sym_size uses which could potentially be removed, but I'll leave that for follow up. 


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

eellison committed:
Add Support For Symbolic SDPA Pattern Matching

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

voznesenskym commented on torch/_inductor/utils.py:
```diff
@@ -576,6 +576,17 @@ def free_symbol_has(index: sympy.Expr, pattern: str):
     return any(pattern in v.name for v in index.free_symbols)
 
 
+def is_symbolic(a: Any) -> bool:
+    return isinstance(a, torch.SymInt) or (
+        isinstance(a, torch.Tensor)
+        and any(is_symbolic(x) for x in itertools.chain(a.size(), a.stride()))
+    )
```
Why not `has_free_symbols` from symbolic_shapes?

eellison replied:
Can use that - this is just a refactoring of existing code

voznesenskym commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -70,7 +72,64 @@ def decompose(decomposition_table):
 py_sym_types = (SymInt, SymFloat, SymBool)
+
+
+class SuppressGuardsAndCheckIfAnyAdded:
```
(1) This should be in shape_env 
(2) Where are you suppressing? by trimming in exit? That isn't sound or right to do
(3) This entirely ignores how produce_guards works
(4) There's already a _suppress_guards_tls

eellison replied:
> That isn't sound or right to do
Can you explain why ? I am admittedly ignorant

> (3) This entirely ignores how produce_guards works

How so ? 

> `_suppress_guards_tls` 

This doesn tdo the same thing.


I can switch to just checking if the expressions are equal, i just would like to know why this doesnt work.

voznesenskym commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -70,7 +72,64 @@ def decompose(decomposition_table):
+    def __hash__(self) -> int:
+        return hash((type(self.sym_obj), hash(self.sym_obj.node.expr)))
+
+    def __eq__(self, value: "SymExprHash") -> bool:
+        assert isinstance(value, SymExprHash)
+
+        sg = SuppressGuardsAndCheckIfAnyAdded(self.sym_obj.node.shape_env)
+        with sg:
+            eq = definitely_true(self.sym_obj == value.sym_obj)
+
+        return eq and not sg.guards_added
```
This is a weird way to do it - I thought we agreed to check the underlying expr, and assume that identical expr means its sound to follow a caching pattern and produce an identical symnode, instead of doing all this guard business?

eellison replied:
Can you explain why the guard business is not sound ? im happy to switch but you havent given any details of why this incorrect.

voznesenskym replied:
1) Design complaint: guards are a notion internal to executing shape_env operations - you have no idea how many other downstream systems are using the view of the guards as they are added. It's very onerous to go from an append-only system to a system that can remove elements. 

2) Correctness complaint: There's a lot of stuff that can happen when evaluating an op that *does not* incur a guard but accumulates some stored entities in a structure that changes which guards are installed. I dont know all the things off the top of my head - but let's look at a simple example. Imagine you run an operation that introduces a new symbol under SuppressGuardsAndCheckIfAnyAdded. A guard will not be added there, but we will have a value in val_to_var forever, as you are not unrolling shape_env state, just trimming guards. All downstream systems will behave as if that symbol is there, when in reality, it was only used for some speculation. The API of SuppressGuardsAndCheckIfAnyAdded does not actually protect you from side effects, it is insufficient to prove that running ops under it does not change any underlying structures.

voznesenskym reviewed (dismissed):
PR does not seem to match what we discussed

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add Support For Symbolic SDPA Pattern Matching"


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Add Support For Symbolic Shapes in Register_replacement, SDPA Pattern Matching"


Many of our pattern matching replacements are specified as a `search_fn` and a `replacment_fn`. The search_fn's are traced out once with static shapes, converted to a pattern, and then matched on every graph compiled with inductor. 
 
The static shape patterns would not match with graphs that are traced out with dynamic shapes because SymInts would be added to the graph as `sym_size` fx nodes which added additional uses and prevented matching. The previous PR partially addresses this by deduping SymInts that are resolvable to graph inputs, as is the calling convention in aot autograd. 

This PR adjusts our matching of the `search_fn` by adding SymInts to the arguments we trace out the search_fn with so that their symint accesses are deduped. Later, if we have a match, we will trace out the replacement graph with the correct Tensors and corresponding symbolic shapes that will get added to the graph.

Note: the replacement patterns will insert sym_size uses which could potentially be removed, but I'll leave that for follow up. 


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Add Support For Symbolic Shapes in Register_replacement, SDPA Pattern Matching"


Many of our pattern matching replacements are specified as a `search_fn` and a `replacment_fn`. The search_fn's are traced out once with static shapes, converted to a pattern, and then matched on every graph compiled with inductor. 
 
The static shape patterns would not match with graphs that are traced out with dynamic shapes because SymInts would be added to the graph as `sym_size` fx nodes which added additional uses and prevented matching. The previous PR partially addresses this by deduping SymInts that are resolvable to graph inputs, as is the calling convention in aot autograd. 

This PR adjusts our matching of the `search_fn` by adding SymInts to the arguments we trace out the search_fn with so that their symint accesses are deduped. Later, if we have a match, we will trace out the replacement graph with the correct Tensors and corresponding symbolic shapes that will get added to the graph.

Note: the replacement patterns will insert sym_size uses which could potentially be removed, but I'll leave that for follow up. 


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Add Support For Symbolic Shapes in Register_replacement, SDPA Pattern Matching"


Many of our pattern matching replacements are specified as a `search_fn` and a `replacment_fn`. The search_fn's are traced out once with static shapes, converted to a pattern, and then matched on every graph compiled with inductor. 
 
The static shape patterns would not match with graphs that are traced out with dynamic shapes because SymInts would be added to the graph as `sym_size` fx nodes which added additional uses and prevented matching. The previous PR partially addresses this by deduping SymInts that are resolvable to graph inputs, as is the calling convention in aot autograd. 

This PR adjusts our matching of the `search_fn` by adding SymInts to the arguments we trace out the search_fn with so that their symint accesses are deduped. Later, if we have a match, we will trace out the replacement graph with the correct Tensors and corresponding symbolic shapes that will get added to the graph.

Note: the replacement patterns will insert sym_size uses which could potentially be removed, but I'll leave that for follow up. 


Fix for https://github.com/pytorch/pytorch/issues/111190.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

ezyang commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -919,15 +929,53 @@ def check_fn(match: Match):
-                )
-                return False
+                    for v in itertools.chain(args[i].shape, args[i].stride()):
+                        if isinstance(v, torch.SymInt) and v not in sym_args:
```
This is highly suspicious. Python will do an equality test when testing for membership, so `v not in sym_args` will effectively install a `!=` guard on every entry in `sym_args`. I can't easily tell what this logic is trying to do in reality though...

eellison replied:
It's gathering all of the unique symints as inputs. as stated below:

```
                # AOT Autograd and make fx will dedupe symbolic shape size
                # accesses of sym ints that appear as inputs
                # We don't want the sym_size uses to interfere with pattern matching
                # so we provide them as inputs.
 ```
